src/Symfony/Component/String/Utf8String.php

```diff
@@ -0,0 +1,249 @@
+        }
+
+        if ('' === $this->string) {
+            return [];
+        }
+
+        $rx = '/(';
+        while (65535 < $length) {
+            $rx .= '.{65535}';
+            $length -= 65535;
+        }
+        $rx .= '.{'.$length.'})/us';
+
+        $str = clone $this;
+        $chunks = [];
+
+        foreach (preg_split($rx, $this->string, -1, PREG_SPLIT_DELIM_CAPTURE | PREG_SPLIT_NO_EMPTY) as $chunk) {
+            $str->string = $chunk;
+            $chunks[] = clone $str;
+        }
+
+        return $chunks;
+    }
+
+    public function endsWith($suffix): bool
+    {
+        if ($suffix instanceof AbstractString) {
+            $suffix = $suffix->string;
+        } elseif (!\is_string($suffix)) {
+            return parent::endsWith($suffix);
+        }
+
+        if ('' === $suffix || !preg_match('//u', $suffix)) {
+            return false;
+        }
+
+        if ($this->ignoreCase) {
+            return preg_match('{'.preg_quote($suffix).'$}iu', $this->string);
+        }
+
+        return \strlen($this->string) - \strlen($suffix) === strrpos($this->string, $suffix);
+    }
+
+    public function equalsTo($string): bool
+    {
+        if ($string instanceof AbstractString) {
+            $string = $string->string;
+        } elseif (!\is_string($string)) {
+            return parent::equalsTo($string);
+        }
+
+        if ($this->ignoreCase) {
+            return mb_strlen($string, 'UTF-8') === mb_strlen($this->string, 'UTF-8') && 0 === mb_stripos($this->string, $string, 'UTF-8');
+        }
+
+        return $string === $this->string;
+    }
+
+    public function indexOf($needle, int $offset = 0): ?int
+    {
+        if ($needle instanceof AbstractString) {
+            $needle = $needle->string;
+        } elseif (!\is_string($needle)) {
+            return parent::indexOf($needle, $offset);
+        }
+
+        if ('' === $needle) {
+            return null;
+        }
+
+        $i = $this->ignoreCase ? mb_stripos($this->string, $needle, $offset, 'UTF-8') : mb_strpos($this->string, $needle, $offset, 'UTF-8');
+
+        return false === $i ? null : $i;
+    }
+
+    public function indexOfLast($needle, int $offset = 0): ?int
+    {
+        if ($needle instanceof AbstractString) {
+            $needle = $needle->string;
+        } elseif (!\is_string($needle)) {
+            return parent::indexOfLast($needle, $offset);
+        }
+
+        if ('' === $needle) {
+            return null;
+        }
+
+        $i = $this->ignoreCase ? mb_strripos($this->string, $needle, $offset, 'UTF-8') : mb_strrpos($this->string, $needle, $offset, 'UTF-8');
+
+        return false === $i ? null : $i;
+    }
+
+    public function length(): int
+    {
+        return mb_strlen($this->string, 'UTF-8');
+    }
+
+    public function prepend(string ...$prefix): AbstractString
+    {
+        $str = clone $this;
+        $str->string = (1 >= \count($prefix) ? $prefix[0] ?? '' : implode('', $prefix)).$this->string;
```

Comment:
The code consistently uses `clone $this` to derivate copies, that should be preserved.

Reply:
@nicolas-grekas yeah, but I don't want to get rid of `clone`, but re-use logic from `append()` inside `prepend()` by appending to empty string and then prepending to original value while assigning to cloned one, like this:

```suggestion
-        $str->string = (1 >= \count($prefix) ? $prefix[0] ?? '' : implode('', $prefix)).$this->string;
+        $str->string = (new static())->append($prefix)->string.$this->string;
```
It's subjective, but I don't like to reuse such code "structures" like `1 >= \count($prefix) ? $prefix[0] ?? '' : implode('', $prefix)`, which is duplicated in those methods. Optionally it could be like:
```
# prepend()
$str->string = $this->normalizeTokens($prefix).$this->string;

# append()
$str->string .= $this->normalizeTokens($suffix);
```
where `normalizeTokens()` (could be named differently of course) would do the "magic" with imploding when there are multiple prefixes/suffixes or just using first one. I know you don't like additional hops, but it really improves readability if well-named, common methods are used in top-level methods. You know what code does and if you need specific information (for example how utf8 strings are determined in `isValidUtf8String()` which I suggested before), you go there. Most of the time you won't need to go deeper, you'll want to see how those top-level methods work.

Reply:
Mmm I see. On my side, I don't like making methods from onliners. I methods should first make sense in an abstract way, then have a body for the implementation. DRY is not the best guide to me. I'd prefer keeping things as is.

Reply:
Yeah, it's up to you. It was only suggestion 🙂